### PR TITLE
feat: [DHIS2-12365] add d2:extractDataMatrixValue

### PR DESCRIPTION
**Pull request by Bekkalizer**
Added support for d2:extractDataMatrixValue(key, dataMatrixText) as described in the epic: https://jira.dhis2.org/browse/DHIS2-12353

The javascript code is based on the backend java rule engine implementation built in this ticket: https://jira.dhis2.org/browse/DHIS2-12364

### CODE REVIEW

**Review by superskip — CHANGES_REQUESTED** [review]
Very easy-to-read code! :+1: 

The solution is a bit inefficient though. In particular it seems like such a wasted effort to build up `dataMap`, only to extract a single value from it in the end and then forget it was ever created. One improvement that could be done directly is to memoize the last `dataMatrix` value  that was given as argument, and reuse `dataMap` if the next `dataMatrix` input value equals the memoized value. This would only be an improvement if we expect the same input multiple times in a row, though.

Another efficiency deficiency lies in how `dataMap` is being built (a part of the procedure currently scans the list of existing keys from top to bottom to find a matching key).

Taking only the task that this code is currently supposed to perform into consideration, the direct (and thus more natural?) approach would be to only search through `dataMatrix` for the key we are interested in, and then extract and return the associated value. The extraction process would in general require solving the same problem as the one I mentioned in the preceding paragraph, i.e., find the key matching the "prefix" of a given string. I'm quite certain you can generate some lookup-table data structure from `gs1Elements` which is better suited for finding the prefix. It should then be used in the function `getApplicationIdentifier`, which you will need in either approach.

**Summing up:**
- You should reimplement `getApplicationIdentifier` employing a better suited data structure, which is pregenerated based on `gs1Elements`.
- Fix the two comments in the code.
- Consider memoizing `dataMatrix` between function calls **or** extract the requested value directly without building `dataMap`.

**Comment by Bekkalizer** [comment]
Implemented getApplicationIdentifier with a slight change to iterate the values instead of the keys.

Built a cache that will keep the dataMaps extracted from the dataMatrix text.

**Review by superskip — APPROVED** [review]
LGTM!
